src/cargo/util/toml.rs

```diff
@@ -543,21 +560,52 @@ impl TomlManifest {
 }
 
 fn validate_library_name(target: &TomlTarget) -> CargoResult<()> {
-    if target.name.trim().is_empty() {
-        Err(human(format!("library target names cannot be empty.")))
-    } else if target.name.contains("-") {
-        Err(human(format!("library target names cannot contain hyphens: {}",
-                          target.name)))
-    } else {
-        Ok(())
+    match target.name {
+        Some(ref name) => {
+            if name.trim().is_empty() {
+                Err(human(format!("library target names cannot be empty.")))
+            } else if name.contains("-") {
+                Err(human(format!("library target names cannot contain hyphens: {}",
+                                  name)))
+            } else {
+                Ok(())
+            }
+        },
+        None => Ok(())
     }
 }
 
 fn validate_binary_name(target: &TomlTarget) -> CargoResult<()> {
-    if target.name.trim().is_empty() {
-        Err(human(format!("binary target names cannot be empty.")))
-    } else {
-        Ok(())
+    match target.name {
+        Some(ref name) => {
+            if name.trim().is_empty() {
+                Err(human(format!("binary target names cannot be empty.")))
+            } else {
+                Ok(())
+            }
+        },
+        None => Err(human(format!("binary target bin.name is required")))
+    }
+}
+
+fn validate_example_name(target: &TomlTarget) -> CargoResult<()> {
+    match target.name {
+        Some(_) => Ok(()),
```

Comment:
Yeah validating example, binary, and test names the same way I think is fine (you can consolidate that here)